cpp/include/raft/detail/serialization.hpp

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <raft/core/handle.hpp>
+#include <raft/core/mdarray.hpp>
+
+#include <rmm/device_uvector.hpp>
+
+#include <type_traits>
+#include <vector>
+
+namespace raft::detail {
+
+template <typename T>
+struct serial;
+
+template <typename T, typename... ContextArgs>
+auto call_serialize(uint8_t* out, const T& obj, ContextArgs&&... args) -> size_t
+{
+  return detail::serial<T>::to_bytes(out, obj, std::forward<ContextArgs>(args)...);
+}
+
+template <typename T, typename... ContextArgs>
+auto call_serialize(const T& obj, ContextArgs&&... args) -> std::vector<uint8_t>
+{
+  std::vector<uint8_t> v(
+    call_serialize<T, ContextArgs...>(nullptr, obj, std::forward<ContextArgs>(args)...));
+  call_serialize<T, ContextArgs...>(v.data(), obj, std::forward<ContextArgs>(args)...);
+  return v;
+}
+
+template <typename T, typename... ContextArgs>
+auto call_deserialize(T* p, const uint8_t* in, ContextArgs&&... args) -> size_t
+{
+  return detail::serial<T>::from_bytes(p, in, std::forward<ContextArgs>(args)...);
+}
+
+template <typename T, typename... ContextArgs>
+auto call_deserialize(T* p, const std::vector<uint8_t>& in, ContextArgs&&... args) -> size_t
+{
+  return call_deserialize<T, ContextArgs...>(p, in.data(), std::forward<ContextArgs>(args)...);
+}
+
+template <typename T, typename... ContextArgs>
+auto call_deserialize(const uint8_t* in, ContextArgs&&... args) -> T
+{
+  union res {
+    T value;
+    explicit res(const uint8_t* in, ContextArgs&&... args)
+    {
+      call_deserialize<T, ContextArgs...>(&value, in, std::forward<ContextArgs>(args)...);
+    }
+    ~res() { value.~T(); }  // NOLINT
+  };
+  // using a union to avoid initialization of T and force copy elision.
+  return res(in, std::forward<ContextArgs>(args)...).value;
+}
+
+template <typename T, typename... ContextArgs>
+auto call_deserialize(const std::vector<uint8_t>& in, ContextArgs&&... args) -> T
```

Comment:
Please see above notes on using `span<const uint8_t>` instead of `const vector<uint8_t>&` -- thanks!

Reply:
Let's keep the related discussion here
https://github.com/rapidsai/raft/pull/770#discussion_r935679007